After saving Order Group order_group_id foreign key is null in orders table. TRUNK-5224

jaks80 opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111 Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-5224

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented:
@jaks80 can you use a more meaningful pull request title as advised at? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

wluyima commented on api/src/test/java/org/openmrs/api/OrderServiceTest.java:
```diff
@@ -3413,4 +3413,39 @@ public void shouldSetTheCorrectSortWeightWhenAddingAnOrderWithANegativePosition(
+                
+		OrderGroup savedOrderGroup = Context.getOrderService().getOrderGroupByUuid(orderGroup.getUuid());
+                
+		Context.flushSession();	
```
Do you need to flush? In theory you should actually be testing OrderGroup.addOrder method

wluyima replied:
With that said, the test should be in OrderGroupTest class since you will be testing OrderGroup.addOrder, if it doesn't exist you need to add it.

jaks80 commented:
@wluyima 
I have changed test method according to your suggestion.
OrderGroup.addOrder method is already there and there were lots of tests for it. But its not a static method as you have suggested. I have added only one line in that method to initialize OrderGroup in Order. 
order.setOrderGroup(this);

wluyima commented on api/src/test/java/org/openmrs/api/OrderServiceTest.java:
```diff
@@ -3413,4 +3413,28 @@ public void shouldSetTheCorrectSortWeightWhenAddingAnOrderWithANegativePosition(
 		secondSavedOrderGroup.addOrder(newOrderWithInvalidPosition, secondSavedOrderGroup.getOrders().size() + 1);
 	}
+        
+        @Test
```
You should be testing `OrderGroup` class and not `OrderService` because the unit of code that has the behavior you are testing is in `OrderGroup` and not `OrderService`, that's the meaning of unit testing.

wluyima replied:
To be more specific, you should be testing the `OrderGroup.addOrder(Order)` method

jaks80 commented:
@wluyima I have moved this test to OrderGroupTest.java. And its testing orderGroup.addOrder(firstOrder);
Thank you.

wluyima commented on api/src/test/java/org/openmrs/OrderGroupTest.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @author User
+ */
+public class OrderGroupTest extends BaseContextSensitiveTest{
```
I don't think test requires any access to the DB, so it should not extend `BaseContextSensitiveTest`

wluyima commented on api/src/test/java/org/openmrs/OrderGroupTest.java:
```diff
@@ -0,0 +1,39 @@
+public class OrderGroupTest extends BaseContextSensitiveTest{
+
+    @Test
+    public void addOrder_shouldAddOrderInOrderGroup() {
```
I think should be renamed to something more intuitive like `addOrder_shouldSetTheOrderGroupOnTheAddedOrder()`

jaks80 commented:
Hi @wluyima I have renamed test Method. Also removed BaseContextSensitiveTest.

wluyima commented on api/src/test/java/org/openmrs/api/OrderServiceTest.java:
```diff
@@ -3412,5 +3412,5 @@ public void shouldSetTheCorrectSortWeightWhenAddingAnOrderWithANegativePosition(
 		expectedException.expect(APIException.class);
 		expectedException.expectMessage("Cannot add a member which is out of range of the list");
 		secondSavedOrderGroup.addOrder(newOrderWithInvalidPosition, secondSavedOrderGroup.getOrders().size() + 1);
-	}
```
Can you revert this format change in this file before I can merge the code please? Sorry! I should have seen this earlier.

wluyima commented:
Thanks! Looks good!

jaks80 commented:
Hi @wluyima 
I have reverted formatting. Also removed some unused imports. Please have a look. Thanks for your patience.

wluyima commented on api/src/test/java/org/openmrs/OrderGroupTest.java:
```diff
@@ -0,0 +1,28 @@
+package org.openmrs;
```
Missing license here,

wluyima commented on api/src/test/java/org/openmrs/OrderGroupTest.java:
```diff
@@ -0,0 +1,28 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
```
You can also get rid of these javadocs since they are not saying much, we try to keep attribution out of the code since the VCS already has it anyways.

wluyima replied:
Sorry that I didn't catch these too earlier

jaks80 commented:
@wluyima Lisence Header added and Java Doc Removed.

wluyima commented:
One more thing, can you squash the 8 commits into 1 please? You should refer to the [pull request tips page](https://wiki.openmrs.org/x/KYOfAw) in the future for best practices when creating pull requests to help have your work merged sooner than later.

jaks80 commented:
Hi @wluyima 
I have badly messed up commits while Squashing. It was messed up so badly I had to delete entire branch and create new one. I have closed this(#2280) pull request. New pull Request is  #2284
